### PR TITLE
refactor optimizer loop logic for manual and automatic optimization

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Here I'm separating

1. automatic optimization which has to loop over the optimizers
2. manual optimization which should NOT loop over optimizers

Previously the two were mashed together under one optimizer loop where in the special case of manual optimizatio, the loop contained a single optimizer for some reason.

By splitting it, we are now able to better understand what the return value for the results look like and simplifies further refactors!

Other connected simplifications:
- renamed `get_optimizer_iterable` -> `get_active_optimizers`
- renamed `num_optimizers` -> `num_active_optimizers`
- decoupled accumulating `batch_outputs` from result processing helper function

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -685,69 +702,79 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
+                    if result:
+                        batch_outputs[opt_idx].append(result.training_step_output_for_epoch_end)
+                        grad_norm_dict = result.get("grad_norm_dict", {})
+            else:
+                result = self.run_batch_split(batch_idx, split_idx, split_batch)
+                if result:
+                    # this if check is required for
+                    # tests/trainer/optimization/test_manual_optimization.py::test_step_with_optimizer_closure_and_accumulated_grad
+                    # TODO: make grad accumulation + manual optimization incompatible to simplify this logic here!
+                    batch_outputs[0].append(result.training_step_output_for_epoch_end)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:720]: Here I'm separating 
1. automatic optimization which has to loop over the optimizers
2. manual optimization which should NOT loop over optimizers

previously the two were mashed together under one optimizer loop where in the special case of manual optimizatio, the loop contained a single optimizer for some reason. 

By splitting it, we are now able to better understand what the return value for the results look like and simplifies further refactors!

**Review by akihironitta — COMMENTED** [review]
How about renaming `train_step_and_backward_closure` to `training_step_and_backward_closure` for consistency?

https://github.com/PyTorchLightning/pytorch-lightning/blob/338cfbca9f00e432cf1d70aa7958f05d3e52ed4a/pytorch_lightning/trainer/training_loop.py#L741
https://github.com/PyTorchLightning/pytorch-lightning/blob/338cfbca9f00e432cf1d70aa7958f05d3e52ed4a/pytorch_lightning/trainer/training_loop.py#L800

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -685,69 +692,74 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
+        # ------------------------------
+        # BACKWARD PASS
+        # ------------------------------
+        # gradient update with accumulated gradients
+        else:
+            if self.trainer.lightning_module.automatic_optimization:
 
-                    else:
-                        result = self.training_step(split_batch, batch_idx, opt_idx, self.trainer.hiddens)
+                def train_step_and_backward_closure():
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:741]: Small nits. Mind integrate the changes from this PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/6212/files#diff-6b21474ed45079f01dfa45ee3b9d40d23efe693c9c005ce897e25384ef425349R692.

Let's replace the closure by a partial there. We need to update the Lightning Optimizer there too:
https://github.com/PyTorchLightning/pytorch-lightning/pull/6212/files#diff-d931a456f290914901d2d1456ab3c85e63b39669d00723d624b988213dccbfdeR213

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py]: I integrated the make_closure with partial here in this PR as you requested. 
I don't understand why the partial call it is useful, because all arguments are supplied so the returned function is not really partial.

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -230,23 +231,30 @@ def _should_add_batch_output_to_epoch_output(self) -> bool:
...
+            A list of tuples (opt_idx, optimizer) of currently active optimizers.
         """
         if not self.trainer.optimizer_frequencies:
             # call training_step once per optimizer
             return list(enumerate(self.trainer.optimizers))
 
-        if batch_idx is None:
-            batch_idx = self.total_batch_idx
-
+        batch_idx = self.total_batch_idx if batch_idx is None else batch_idx
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:251]:
```suggestion
        batch_idx = batch_idx or self.total_batch_idx
```

File: tests/deprecated_api/test_remove_1-5.py [hunk 4]
```diff
@@ -196,7 +196,7 @@ def __init__(self):
             self.automatic_optimization = False
 
         def training_step(self, batch, batch_idx, optimizer_idx):
-            assert optimizer_idx is not None
+            assert optimizer_idx is None
```
Comment (carmocca) [inline tests/deprecated_api/test_remove_1-5.py:199]: why has this changed?

File: pytorch_lightning/trainer/training_loop.py [hunk 5]
```diff
@@ -685,71 +693,86 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
+            if self.trainer.lightning_module.automatic_optimization:
+                self.optimizer_step(optimizer, opt_idx, batch_idx, closure)
+            else:
+                result = self.training_step(split_batch, batch_idx, opt_idx, self.trainer.hiddens)
 
-                    # update running loss + reset accumulated loss
-                    self.update_running_loss(result.loss)
+            if not result:
+                # user decided to skip optimization
+                # make sure to zero grad.
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:747]: This comment is outdated. It comes from https://github.com/PyTorchLightning/pytorch-lightning/pull/4485/files#diff-6b21474ed45079f01dfa45ee3b9d40d23efe693c9c005ce897e25384ef425349R701-R702

```suggestion
```

File: tests/deprecated_api/test_remove_1-5.py (hunk 4 above)
Comment (awaelchli) [reply tests/deprecated_api/test_remove_1-5.py:199]: Here, before, optimizer_idx=0 which only exists for backward compatibility when someone has the optimizer_idx argument in the signature of training_step. The value is arbitrary and has no meaning in manual optimization.

Why did it change?
In the above refactor, I grouped the optimization step under one function which is called either

`self.run_optimization(batch_idx, split_idx, split_batch, opt_idx, optimizer)`
in auto optimization 

or 

`self.run_optimization(batch_idx, split_idx, split_batch, opt_idx=None, optimizer=None)`
in manual optimization.

I can change it back to `opt_idx=0` but I find that suboptimal, since referring to the first optimizer is wrong.

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:251]: batch_idx can be 0 too. For sanity I recommend not omitting the None check.

File: pytorch_lightning/trainer/training_loop.py [hunk 6]
```diff
@@ -685,71 +693,85 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
-                        result = self.training_step(split_batch, batch_idx, opt_idx, self.trainer.hiddens)
+            # -------------------
+            # calculate loss (train step + train step end)
+            # -------------------
 
-                    if not result:
-                        # user decided to skip optimization
-                        # make sure to zero grad.
-                        continue
+            # automatic_optimization=True: perform dpp sync only when performing optimizer_step
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:730]:
```suggestion
            # automatic_optimization=True: perform ddp sync only when performing optimizer_step
```

File: pytorch_lightning/trainer/training_loop.py [hunk 7]
```diff
@@ -685,71 +693,85 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
+            if self.trainer.lightning_module.automatic_optimization:
+                self.optimizer_step(optimizer, opt_idx, batch_idx, closure)
+            else:
+                result = self.training_step(split_batch, batch_idx, opt_idx, self.trainer.hiddens)
 
-                    # update running loss + reset accumulated loss
-                    self.update_running_loss(result.loss)
+            if not result:
+                # user decided to skip optimization
+                return result
```
Comment (ananthsub) [inline pytorch_lightning/trainer/training_loop.py:747]: should this be an explicit None check? is returning False or 0 from training step allowed?

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:747]: it returns an AttributeDict, which can be empty and therefore I don't check for None.

File: pytorch_lightning/trainer/training_loop.py [hunk 8]
```diff
@@ -230,23 +231,30 @@ def _should_add_batch_output_to_epoch_output(self) -> bool:
 
         return False
 
-    def get_optimizers_iterable(self, batch_idx=None):
+    def _should_skip_optimizer(self, opt_idx: int, batch_idx: Optional[int] = None) -> bool:
+        """ Determine if the optimizer should be skipped based on desired frequencies. """
+        active_indices = [idx for (idx, _) in self.get_active_optimizers(batch_idx)]
+        return opt_idx not in active_indices
+
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:238]: nowhere, will remove.

File: pytorch_lightning/trainer/training_loop.py [hunk 9]
```diff
@@ -230,23 +231,30 @@ def _should_add_batch_output_to_epoch_output(self) -> bool:
...
+    def get_active_optimizers(self, batch_idx: Optional[int] = None) -> List[Tuple[int, Optimizer]]:
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:239]: I just renamed the existing method

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:234]: trainer calls it, finetune callback calls it. 
aim for this PR is not to improve all parts of this train loop. one step at a time.

File: pytorch_lightning/trainer/training_loop.py [hunk 10]
```diff
@@ -78,9 +80,8 @@ def __init__(
             self.trainer.num_sanity_val_steps = num_sanity_val_steps
 
     @property
-    def num_optimizers(self):
-        num_optimizers = len(self.get_optimizers_iterable())
-        return num_optimizers
+    def num_active_optimizers(self):
+        return len(self.get_active_optimizers())
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:84]: could be, does not matter here.

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py [hunk 11]
```diff
@@ -685,71 +688,85 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
-                    # -------------------
-
-                    # automatic_optimization=True: perform dpp sync only when performing optimizer_step
-                    # automatic_optimization=False: don't block synchronization here
-                    with self.block_ddp_sync_behaviour():
-                        result = self.training_step_and_backward(
-                            split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens
-                        )
+            if self.trainer.lightning_module.automatic_optimization:
+                for opt_idx, optimizer in self.get_active_optimizers(batch_idx):
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:692]: Personal, get_active_optimizers make it sound there is inactive optimizers. Might be confusing to readers.

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:696]: There are inactive optimizers. For example with two optimizers where one has frequency 1 and the other frequency 2. 
Then every other training step we only have 1 active optimizer.

File: pytorch_lightning/core/optimizer.py [hunk 12]
```diff
@@ -207,7 +207,7 @@ def closure_dis():
             profiler_name = "closure_{self._optimizer_idx}"
             closure = do_nothing_closure
         else:
-            if not isinstance(closure, types.FunctionType):
+            if not isinstance(closure, Callable):
```
Comment (justusschock) [inline pytorch_lightning/core/optimizer.py:210]: those checks don't work for all cases. Better replace it with `callable(closure)`
```suggestion
            if not callable(closure):
```

File: pytorch_lightning/trainer/training_loop.py (hunk 7 above)
Comment (justusschock) [reply pytorch_lightning/trainer/training_loop.py:746]: So if users return None in train step, we return an empty attributedict instead?

File: tests/deprecated_api/test_remove_1-5.py (hunk 4 above)
Comment (justusschock) [reply tests/deprecated_api/test_remove_1-5.py:199]: Actually I would change It back, since this means the same code won't run when you switch from 1 to to 2 optims if you check that value ( I get your point though, but also when you only have one optimal isn't it always the first (and only)?)

**Review by justusschock — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-5.py (hunk 4 above)
Comment (awaelchli) [reply tests/deprecated_api/test_remove_1-5.py:199]: It would anyway not run, because we don't iterate over optimizers in manual optimization and the optimizer_idx argument is an artifact from deprecating that argument from training_step. 

If the user wants multiple optimizers, they have to do:

```python
opt1, opt2 = self.optimizers()

self.manual_backward(loss1)
opt1.step()
self.manual_backward(loss2)
opt2.step()
```

Otherwise nothing works in manual optimization.

File: pytorch_lightning/trainer/training_loop.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:749]: yes, internally. 
makes logic for managing these results slightly easier.
